### ISSUE TITLE
Implementing/Training PointNet for alpha_s with various epochs and folds and finding the best learning rate

### ISSUE THREAD

**Issue by hmehryar** [hmehryar/Hm.JetscapeMl#309]
model: PointNet
Classifying param: alpha_s
epochs: 10, 20, 32, 50
fold:1
learning rate: 0.001

**Comment by hmehryar** [comment]
it seems that something is missing in my calculation, I'm gonna start over

**Comment by hmehryar** [comment]
I think I don't need to split the original dataset_x anymore, cuz I'm just working with dataset_x_points, so I reimplemented it without having the dataset_x

**Comment by hmehryar** [comment]
in the last commit 

1. added callbacks to the model
2. made an indivisual function for train
3. saved train history 
4. saved accuracy/loss diagram for epochs
5. saved confusion matrix /accuracy for test

**Comment by hmehryar** [comment]
Notes: Date 5/12/24 Day 73: @hmehryar 
- working on making the classifier to treat the dataset file just like the original PointNet code which works with a better performance on GPU
- re-studying  the original pointnet code on Colab
- modification on pointNEt_alpha_S implmenetation
- > - changing the final layer activation function to softmax
      -  changing the accuracy/loss to sparse categorical accuracy/loss
      - it seems by making it categorical, it doesn't end up in early stopping
      - witnessed a 40% accuracy on epoch 2/20
      - without Augmentation and with sparse categorical loss/accuracy, and softmax as activation early stopping on epoch 8/20

**Comment by hmehryar** [comment]
@hmehryar  Notes: Implementation: PointNet_alpha_s 1k_1fold_20epochs on HmSrv1 

- implementing slice function
- implementing shuffle/augmentation  and converting to train and validation
- implementation train_model_with_callback with TensorFlow dataset (with slice and augmentation)
- implementing plot training history to work with sparse categorical accuracy and validation
- implementing evaluate_model to work with slice TensorFlow dataset 
- moved augment function to pointNEt.py
- cleaned up the package from notebook headers
- delete extra comment code from pointnet.phy
- delete extra notebook from alpha_s folder/namespace
-

**Comment by hmehryar** [comment]
regarding facing early stopping, I'm gonna implement the scaling dataset images.

**Comment by hmehryar** [comment]
![jetml_pointnet_classification_alpha_s_0 2_0 3_0 4_size_1000_epochs_32_fold_1_accuracy_loss](https://github.com/hmehryar/Hm.JetscapeMl/assets/9429125/b310f3f7-1395-4078-bbdc-f429ec38ecb3)
[jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000_epochs_32_fold_1_results_kfold.txt](https://github.com/hmehryar/Hm.JetscapeMl/files/15312745/jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000_epochs_32_fold_1_results_kfold.txt)

**Comment by hmehryar** [comment]
regarding facing early stopping, I'm gonna implement the scaling dataset images.

**Comment by hmehryar** [comment]
![jetml_pointnet_classification_alpha_s_0 2_0 3_0 4_size_1000_epochs_32_fold_1_accuracy_loss](https://github.com/hmehryar/Hm.JetscapeMl/assets/9429125/473c4da5-62ce-4dd3-9075-8c7aa9ed5c1e)
[jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000_epochs_32_fold_1_results_kfold.txt](https://github.com/hmehryar/Hm.JetscapeMl/files/15313349/jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000_epochs_32_fold_1_results_kfold.txt)

**Comment by hmehryar** [comment]
I'm going to run this version for 100k and 1000k on grid if i can get my hand on v100 now, hopefully i get 70% and 90% respectively on val_accuracy

**Comment by hmehryar** [comment]
pointnet_alpha_s_size_10k_fold_1_epochs_100_colab_GPU_early_22_best_val_acc_499

![jetml_pointnet_classification_alpha_s_0 2_0 3_0 4_size_10000_epochs_100_fold_1_accuracy_loss](https://github.com/hmehryar/Hm.JetscapeMl/assets/9429125/e6a6a555-ef0f-4585-b877-07b2a8280d20)
[jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_10000_epochs_100_fold_1_results_kfold.txt](https://github.com/hmehryar/Hm.JetscapeMl/files/15337556/jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_10000_epochs_100_fold_1_results_kfold.txt)

**Comment by hmehryar** [comment]
pointnet_alpha_s_size_100k_fold_1_epochs_100_colab_GPU_early_13_best_val_acc_488_test_acc_329
![jetml_pointnet_classification_alpha_s_0 2_0 3_0 4_size_100000_epochs_100_fold_1_accuracy_loss](https://github.com/hmehryar/Hm.JetscapeMl/assets/9429125/7b4f80fc-b7c8-4e62-8f8e-1a9a6e7018a9)
[jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_100000_epochs_100_fold_1_results_kfold.txt](https://github.com/hmehryar/Hm.JetscapeMl/files/15337557/jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_100000_epochs_100_fold_1_results_kfold.txt)

**Comment by hmehryar** [comment]
I'm going to close this issue, with this results, and open another issue for hyper parameter tuning the PointNet classifier for alpha_s

**Comment by hmehryar** [comment]
Trying to run the pointnet_alpha_s_1000k_1f_100epochs on Grid_with GPU

**Comment by hmehryar** [comment]
Grid CUDA couldn't run 1000k dataset by GPU computation on TensorData  regarding  the memory constraints
ends up to segmentation error
I started optimizing the code by deleting non-necessary variables/objects, it wasn't good enough
did more research about it, 
So I implemented a technique to do training stuff on on CPU and just put the training on GPU.
and it worked

**Comment by hmehryar** [comment]
pointnet_alpha_s_size_1000k_fold_1_epochs_100_grid_GPU_early_13_patience_5_best_val_acc_488_test_acc_334
![jetml_pointnet_classification_alpha_s_0 2_0 3_0 4_size_1000000_epochs_100_fold_1_accuracy_loss](https://github.com/hmehryar/Hm.JetscapeMl/assets/9429125/1c3193c1-4e1d-4ffe-8d3b-2fd0a825d8b4)
[jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000000_epochs_100_fold_1_results_kfold.txt](https://github.com/hmehryar/Hm.JetscapeMl/files/15366613/jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000000_epochs_100_fold_1_results_kfold.txt)

**Comment by hmehryar** [comment]
pointnet_alpha_s_size_1000k_fold_1_epochs_100_grid_GPU_early_42_patience_40_best_val_acc_472_test_acc_331
![jetml_pointnet_classification_alpha_s_0 2_0 3_0 4_size_1000000_epochs_100_fold_1_accuracy_loss](https://github.com/hmehryar/Hm.JetscapeMl/assets/9429125/85de4ef0-8139-4c25-b981-39e913bfa3a6)
[jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000000_epochs_100_fold_1_results_kfold.txt](https://github.com/hmehryar/Hm.JetscapeMl/files/15366616/jetml_pointnet_classification_alpha_s_0.2_0.3_0.4_size_1000000_epochs_100_fold_1_results_kfold.txt)

**Comment by hmehryar** [comment]
GRID configuration
27000600       gpu R bash                           gy4065    1   16    199G voh5                     v100

 nvidia-smi
Sat May 18 15:43:15 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-SXM2-16GB           On  | 00000000:1A:00.0 Off |                    0 |
| N/A   37C    P0              40W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2-16GB           On  | 00000000:1D:00.0 Off |                    0 |
| N/A   31C    P0              38W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2-16GB           On  | 00000000:1E:00.0 Off |                    0 |
| N/A   35C    P0              39W / 300W |      0MiB / 16384MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
